FQDN: transition to asynchronous IPCache APIs

squeed opened the pull request:
This PR transitions the FQDN policy subsystem off of the synchronous ipcache APIs.

This design keeps the 3 major components:
- **SelectorCache**: Maps label selectors to a set of numeric identities. This also distributes the incremental identity change stream to its users, namely policies and endpoints
- **DNSCache**: Maintains a mapping of name -> ip and ip -> name. Unchanged.
- **NameManager**: Watches (name -> ip) changestream, tracks which selectors need to be updated accordingly.

Previous to this PR, the NameManager would allocate numeric identities based on DNS requests / responses from pods. It would then distribute identities to all relevant FQDN selectors in the SelectorCache. The SelectorCache selectors would then be responsible for de-allocating the identities upon deletion.

This PR changes this slightly:

1. The NameManager now allocates identities asynchronously, via the ipcache metadata layer. (It also manages deletion, and thus de-allocates them as well).
2. A FQDNSelector in the SelectorCache is now a bag of labels. Specifically, it is a bag of `cidr:xxxx/32` labels. These labels are updated dynamically by the NameManager. This is the same logical flow as before, where updates were pushed to cached selectors by the NameManager -- just the content of the update is different

The biggest non-obvious change is with locking. Previously, the NameManager needed much of the system unlocked in order to allocate identities. Now that this is all asynchronous, the only interesting locks are SelectorCache and NameManager. As long as they are acquired in the same order (NameManager before SelectorCache), then we are good. This removes any dependency on Endpoint or IdentityAllocator locks.

This PR has a bit of a potential performance improvement, in that if two selectors select the same IPs, then we will shortcut the identity allocation / ipcache upsert the second time.

### Tradeoffs

~There is one significant downside to this new approach. The ipcache asynchronous loop cannot make progress until the k8s caches are synchronized. That means that any DNS responses that detect new IPs may cause policy drops until the caches are synced. I believe we can improve this (in a follow-up PR) by limiting the set of resources that need to be synchronized.~

**EDIT:** this is wrong. We don't actually cause any additional delay for a complicated reason. This PR doesn't change the behavior at all!

(note to reviewers: I did my best to split up commits. However, there was one big commit that was pretty unavoidable. Apologies in advance).

TODO:
- [x] Update docblocks.
- [ ] way more unit tests
- [x] block policy response until the ipcache has pushed changes, _but only if needed_

Fixes: #28930

squeed committed:
fqdn: move GC, cleanup, restore functions to NameManager

Rather than having these functions in the top-level daemon, move them to
the NameManager. This commit has no functional changes.

Future commits will add additional functionality within these functions.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
fqdn: (cleanup) use sets.Set instead of maps

Under the hood it's the same, but it gets rid of some visual clutter.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
fqdn: create an ipcache metadata entry for all FQDN IPs

This refactors the fqdn NameManager to create an ipcache entry for all IPs
known by the FQDN subsystem. It does so by intercepting all calls that
can lead to addition and removal of IPs.

Right now, this metadata entry is not immediately used; when the fqdn
selectors stop allocating identities manually, this will be important.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
policy: split out selectorcache.go

This is a noop change that just moves some functions and definitions
around. No code has been changed.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
fqdn: move to asynchronous ipcache APIs

Rather than tracking identities directly, this changes the fqdn selector
to be a "bag of labels". The, the NameManager pushes desired IPs
directly down in to the FQDN selector. Since the identities will be
allocated via the ipcache (and pushed directly to the selectors), we can
completely get out of the identity management game.

Everything else is just bookkeeping and (lots and lots and lots) of code
cleanup.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

christarazi commented on pkg/fqdn/name_manager.go:
```diff
@@ -257,3 +257,28 @@ func (n *NameManager) updateIPsForName(lookupTime time.Time, dnsName string, new
+	for _, ip := range ips {
+		prefixes = append(prefixes, netip.PrefixFrom(ip, ip.BitLen()))
+	}
+	n.config.IPCache.UpsertPrefixes(prefixes, source.Generated, ipcacheResource)
```
Shouldn't this wait on the channel now that `UpsertPrefixes` returns one?

squeed replied:
That’s the plan. However, we can short-cut in the case where all the IPs already have identities. I need to find the right place to do this.

squeed replied:
Also, we can drop the NameManager lock before waiting on the channel. This just needs a bit of thought.

squeed replied:
OK, I've figured out the concurrency story reasonably well. I'll describe it more in the PR description.

squeed committed:
fqdn: cleanup unused code

we no longer need the multi-ip-mutex mechanism, now that the ipcache
handles identity allocation directly. So, remove it.

Since we're not allocating identities anymore, we don't need a
reference-counting allocator for tests.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
policy/selectorcache: invert identitySelector interface

Now that fqdn and label selectors look very similar, we can simplify
this code a lot by a small refactor. Now, the only "business" logic is a
selector-specific `matches()` method.

This does not change any behavior.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed commented on pkg/policy/selectorcache.go:
```diff
@@ -748,28 +608,29 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted cache.IdentityCache, wg
 			case *fqdnSelector:
-				// This is a no-op right now. We don't encode in the identities
-				// which FQDNs they correspond to.
+				for numericID := range deleted {
```
note to reviewers: this duplicated code is removed in a subsequent commit.

squeed commented on pkg/policy/selectorcache_selector.go:
```diff
@@ -0,0 +1,356 @@
+// Because the selector exposed to the user is used as a map key, it must always
+// be passed to the user as a pointer to the actual implementation type.
+// For this reason 'notifyUsers' must be implemented by each type separately.
+type identitySelector struct {
```
note to self: this comment is no longer correct; need to update.

christarazi commented on pkg/ipcache/metadata.go:
```diff
@@ -80,6 +80,13 @@ type metadata struct {
+	// but have already allocated a new done channel.
+	// The length of this is theoretically unbound, but in reality it will
+	// never have more than 2 entries. It will always have at least one entry.
+	queuedPrefixesDone []chan struct{}
```
I'm having a bit of a hard time following this even though I can tell you've put effort into trying to explain it. I am wondering whether it's simpler to have a revision number instead that's incremented each time a prefix is processed by ipcache. Instead of a channel, the next revision number can be returned and a `WaitForRevision` function can be used to wait for the ipcache to implement the revision. I think this will also simplify a lot of the calling code and APIs that have to now handle / return channels, i.e. ensuring their closes in certain cases, etc, and just eliminates the need for all that complexity.

squeed replied:
Oh sure, that's a reasonable idea. This is actually the right case to use `sync.Cond{}`.

squeed replied:
@christarazi I changed this to use a revision number; PTAL

christarazi commented on pkg/fqdn/config.go:
```diff
@@ -26,4 +26,13 @@ type Config struct {
 	// sets of IPs.
 	UpdateSelectors func(ctx context.Context, selectorsWithIPs map[api.FQDNSelector][]net.IP, selectorsWithoutIPs []api.FQDNSelector) (*sync.WaitGroup, []*identity.Identity, map[netip.Prefix]*identity.Identity, error)
+
+	// GetEndpointsDNSInfo is a function that returns a list of interesting structures within an Endpoint
```
We should probably clarify what's "interesting" what these structures :)

squeed replied:
Fair enough ;-)

christarazi commented on pkg/fqdn/name_manager_gc.go:
Very happy to see this code finally get refactored into its own pkg, rather than being awkwardly in the daemon.

christarazi commented on pkg/fqdn/name_manager.go:
```diff
@@ -266,3 +266,29 @@ func (n *NameManager) updateIPsForName(lookupTime time.Time, dnsName string, new
 	return (len(cacheIPs) == 0 && len(sortedNewIPs) == 0) || !ip.SortedIPListsAreEqual(sortedNewIPs, cacheIPs)
 }
+
+var ipcacheResource = ipcacheTypes.NewResourceID(ipcacheTypes.ResourceKindDaemon, "", "fqdn-name-manager")
```
Nit: Given this code is outside of the daemon, we could have `ResourceKindFQDN` and the name instead be `name-manager`, but this is nit picky. I leave this comment because all usages of `ResourceKindDaemon` are within the daemon pkg.

squeed replied:
Right, I thought about this for a while; all of the other resource kinds represent some kind of external data source, (kvstore, crd, etc). So I kept `daemon` in this case.

christarazi commented on pkg/fqdn/name_manager_gc.go:
```diff
@@ -44,6 +44,10 @@ var (
 
+	// Take a snapshot of the *entire* reverse cache, so we can compute the set of
+	// IPs that have been completely removed and safely delete their metadata.
+	maybeStaleIPs := n.cache.dumpReverse()
```
Why does the set need to be reversed in GC?

joestringer replied:
I think this is just unclear naming: The `NameManager` stores name->IP and IP->name mappings via `DNSCache`. The latter is named `reverse` internally. The `dumpReverse()` call here is pulling the set of IPs from the `NameManager`. Probably this would be clearer with a name like `n.cache.DumpIPs()` or similar.

squeed replied:
Agreed, the naming is unclear. All we want is the set of in-use IPs. I'll tighten this up.

squeed replied:
DumpIPs makes sense; will update.

christarazi commented on daemon/cmd/fqdn.go:
```diff
@@ -61,97 +61,61 @@ const (
-			"fqdnSelectors": selectorsWithoutIPs,
-		}).Debug("removing all identities from FQDN selectors")
-		d.policy.GetSelectorCache().RemoveIdentitiesFQDNSelectors(selectorsWithoutIPs, notifyWg)
+	// UpdatePolicyMaps consumes notifyWG, and returns its own WaitGroup
+	// that is Done() when all endpoints have pushed their incremental changes
+	// down in to their bpf PolicyMap.
+	if updateResult&policy.UpdateResultUpdatePolicyMaps > 0 {
+		logger.Debug("FQDN selector update requires UpdatePolicyMaps.")
+		wg = d.endpointManager.UpdatePolicyMaps(ctx, notifyWg)
+	} else {
```
As mentioned in an offline sync, this `UpdatePolicyMaps` shouldn't necessarily be required. Before this PR, it seems like we have a double call to `UpdatePolicyMaps` call (endpoint regenerations), one from the FQDN selector layer and one from the ipcache. This PR just keeps that structure, which is ok for now. Let's put a comment here and maybe file an issue to reference for removing the double endpoint regeneration so that we can tackle it in the future. For posterity, the idea discussed was to plumb extra metadata down to the ipcache so that it can call `UpdatePolicyMaps` to update endpoints that a selector now has a new identity relevant to them. Today, ipcache no-ops in this case because it relies on the FQDN selectorcache layer. (I believe I got this right, but please correct me if I got it wrong.)

joestringer replied:
([parent thread](https://github.com/cilium/cilium/pull/29036#discussion_r1395144280))

I thought about this a bit as well, and came to the conclusion that we _do_ want this here as well.

Consider there are a couple of paths for learning FQDN information:
1. An endpoint makes a request for a brand new FQDN and learns the brand new IP. All identity allocation, policy plumbing, ipcache plumbing must occur to allow the endpoint to connect via `toFQDNs` policy.
2. Another endpoint makes a request for an FQDN that maps to the same IP. FQDN may or may not be new. The Security Identity for the IP is already allocated. Policy is not yet plumbed for _this_ endpoint. IPCache plumbing not required as it was already done previously.

This code can gracefully handle case (2). The `UpdateFQDNSelector()` case above will do `setSelectorIPs()` to update the label selectors and then re-evaluate the matches against the `SelectorCache`, which may identify existing identities that are now selected by the `FQDNSelector`. That function will then `notifyUsers()` to enqueue the policymap updates for the current endpoint, but not yet implement the policies into the datapath bpf maps. This `UpdatePolicyMaps()` call then pushes such updates into the datapath for this case.

I would assert that removing this case would then break case (2), because the other path requires the IP to be new in order to allocate a new identity and then propagate the policymap updates through.

christarazi replied:
Correct, if we removed `UpdatePolicyMaps` here we would break case (2). To be clear, I was not suggesting that we do that in this PR. If we introduce a mechanism to trigger the ipcache to call `UpdatePolicyMaps` without having a new IP, then technically we could remove this extra call to `UpdatePolicyMaps`. Functionally there's no impact when moving this call to `UpdatePolicyMaps` from here to the ipcache. But what it does do is centralize the logic for triggering the update of policymaps into the ipcache, so that other subsystems don't need to do it themselves.

That's what I meant by:

> For posterity, the idea discussed was to plumb extra metadata down to the ipcache so that it can call UpdatePolicyMaps to update endpoints that a selector now has a new identity relevant to them. Today, ipcache no-ops in this case because it relies on the FQDN selectorcache layer. 

If we feel like there's not much benefit to that, I'm open to that viewpoint. In general, I am interested in exploring the idea to move as much ipcache-related logic into ipcache, rather than potentially complicating call sites with those details. We have been trending in the direction of ipcache becoming the central point for endpoints/policies, so I see it as a natural extension of that direction.

joestringer replied:
:+1: thanks for clarifying. I agree.

(incidentally, this case (2) would make for an interesting lifecycle test... how does this new approach handle updates to ipcache? If multiple endpoints make DNS lookups for the same name, that will result in a single ipcache entry update right? Will that entry remain live until TTL expires and both endpoints stop using that name? Failing to keep the ipcache entry alive for the full zombie period could cause packet loss. In other words, the `RemoveMetadata()` call should only occur for the same ipcache resourceID once all namemanager usage of the corresponding IP has been garbage collected.)

christarazi replied:
Good question. From my reading of the code, the `RemoveMetadata()` is called within `maybeRemoveMetadataLocked()` of the NameManager during the GC routine.

The following steps occur in the GC logic (`(n *NameManager) GC()`):

1. Collect list of `maybeStaleIPs` which is the current, full dump of the DNS cache as IPs
2. Mark all entries under `ep.DNSHistory` for each endpoint
3. Mark all entries under `ep.DNSZombies` for each endpoint
4. Collate list of all alive entries from steps (2) and (3) and **replace** the global DNS cache with the new set of only alive entries
5. Skip each entry from `maybeStaleIPs` if it exists in the new DNS cache, call `RemoveMetadata()` on the ones that don't exist in the DNS cache

From the above high level steps, I don't believe there's a possibility to prematurely delete an entry that would result in a deleted ipcache entry while the DNS entry is still a zombie.

christarazi replied:
(Answering your general point, I do think lifecycle tests like this would greatly beneficial to have in the codebase if we don't already.)

squeed replied:
The policy engine has two sources of incremental updates (IPCache and NameManager), so it has two separate callers of UpdatePolicyMaps. In fact, the ipcache needs to call UpdatePolicyMaps twice, since it needs a strict (add, commit, delete) ordering.

Chris is right that, in the event cases 1 and 2 are *both* relevant to an FQDN update, we will waste some time doing 3 calls to `UpdatePolicyMaps()` when only 2 are required. However, the FQDN and ipcache flows are sufficiently distinct from one another, so it's not easy to coalesce these calls down. Especially because all 3 callers require the update to be functionally synchronous.

One easy change would be to put UpdatePolicyMaps behind a controller / trigger, and coalesce pending calls. Otherwise, we need to start thinking about a more streamlined design for incremental updates.

However, I'd like to understand if this is has a real-world performance implication before spending a lot of time thinking about this. My intuition is that, most of the time, only an ipcache or fqdn incremental update will be relevant. The case where a FQDN response both allocates a new IP *and* newly-selects an existing should be fairly infrequent.

christarazi commented on daemon/cmd/fqdn.go:
```diff
@@ -408,54 +357,6 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
-		// first, G2 executes T4 also as a no-op and releases the msg back to the
-		// pod at T5 before G1 would at T6.
-		mutexAcquireStart := time.Now()
-		mutexes := d.dnsProxyContext.getMutexesForResponseIPs(responseIPs)
-		for _, m := range mutexes {
-			d.dnsProxyContext.responseMutexes[m].Lock()
-			defer d.dnsProxyContext.responseMutexes[m].Unlock()
-		}
-		if d := time.Since(mutexAcquireStart); d >= option.Config.DNSProxyLockTimeout {
```
We should also deprecate `DNSProxyLockTimeout` and `DNSProxyLockCount` flags as well, since this code is being removed.

christarazi reviewed (commented):
This is great work! Very happy to see us paying down a decent chunk of the FQDN tech debt.

I went for a high-level review pass. I did not review in detail all of the refactors on the assumption that most of the code is not new and changed in slight ways like NameManager operating on IPs instead of identities. However, I did pick out a few things that should be addressed and / or clarified. I note that you left a TODO item for unit tests, so wondering if you are planning to add those in this PR. Overall, the direction of the PR LGTM.

Regarding

> There is one significant downside to this new approach. The ipcache asynchronous loop cannot make progress until the k8s caches are synchronized. That means that any DNS responses that detect new IPs may cause policy drops until the caches are synced. I believe we can improve this (in a follow-up PR) by limiting the set of resources that need to be synchronized.

I'm a bit concerned that we ship this PR without the followup and now there's arguably a regression in FQDN policy in v1.15. I don't think we can ship this without the fix, unfortunately.

joestringer commented on pkg/fqdn/name_manager.go:
```diff
@@ -151,43 +163,38 @@ func (n *NameManager) UpdateGenerateDNS(ctx context.Context, lookupTime time.Tim
 
-	return n.config.UpdateSelectors(ctx, selectorIPMapping, namesMissingIPs)
+	// Update SelectorCache selectors and push changes down in to BPF.
+	return n.config.UpdateSelectors(ctx, selectorIPMapping, ipcacheDoneChan)
```
> > @squeed:
> > There is one significant downside to this new approach. The ipcache asynchronous 
loop cannot make progress until the k8s caches are synchronized. That means that any DNS responses that detect new IPs may cause policy drops until the caches are synced. I believe we can improve this (in a follow-up PR) by limiting the set of resources that need to be synchronized.
> 
> @christarazi:
> I'm a bit concerned that we ship this PR without the followup and now there's arguably a regression in FQDN policy in v1.15. I don't think we can ship this without the fix, unfortunately.

[Starting a thread for this discussion point so it's not spread out throughout the PR]

Arguably this `UpdateGenerateDNS()` call is where the delay gets implicitly applied, because the `updateDNSIPs()` call above will allocate the channel that incurs the delay. The channel will only be closed after ipcache begins making progress. This call to `updateSelectors()` mixes that channel into the resulting  `WaitGroup` which will be returned to the caller. The resulting `WaitGroup` induces the delay on DNS responses in order to allow the policy calculations to catch up, up in `notifyOnDNSMsg()`.

Something to note is that currently the FQDN direct identity allocation approach to plumbing ipcache simply _ignores_ other sources of information about the IPs in the traditional way that ipcache was plumbed. This creates inconsistencies in the ipcache plumbing that over time is causing some bugs. At some level, forcing the fqdn subsystem to synchronize with other sources of information is about plugging that gap and ensuring that ipcache entries are generated consistently regardless of how we learn about IP<->information associations.

> I believe we can improve this (in a follow-up PR) by limiting the set of resources that need to be synchronized.

I'm all for optimizing this, but the ipcache needs to wait for all policies to synchronize (to implement CIDR rules), CiliumIdentities, CEP/CES to map to IPs, and nodes for encryption security parameter indexes (SPI). At higher scale with many resources, this will simply take a longer time. As such I'm not sure that this is the actual solution. 

If we break down the DNS handling behaviour into three stages, we have:
1. Agent is down. DNS responses are dropped.
2. Agent is up, DNS proxy is up, ipcache is uninitialized. DNS responses are handled but with a delay of `FQDNProxyResponseMaxDelay`. Policy plumbing is delayed until ipcache initialization.
3. Agent is up, DNS proxy is up, ipcache is up. Policy is plumbed and DNS responses are delayed until the network is ready to handle the corresponding connections.

One way to debate this is whether it's better to be in stage (1) or stage (2). We could defer starting the DNS proxy until k8s sync and then (1) would be longer and (2) would be shorter. Currently with this implementation, stage (2) begins a bit earlier. In this case, the app may receive DNS responses and begin connecting to IPs that the network is not ready to handle. Given TCP retries in the kernel, for many apps this will be resilient, just with some delay in connection time during agent outage.

christarazi replied:
It seems to be that minimizing (1) as much as possible is probably the best thing we can do. Especially considering that if Cilium has restarted, then previous DNS proxy state could have been reloaded which could allow apps to continue to work without much disruption. Whereas if (1) is not minimized, even if DNS proxy state was restored, it would still not handle DNS traffic.

joestringer replied:
I agree that minimizing packet drops as much as possible is best, but under high scale clusters with high kube-apiserver load during agent restarts or Cilium upgrades, I could see (2) taking extended periods of time as well. I'm not sure whether it's ideal for Cilium to be handling such traffic suboptimally during that period (incurring delays and potentially even failing to pick up additional DNS packets because all the DNS-handling goroutines are stuck waiting). Architecturally speaking this is probably where an external DNS proxy component could fill the gap, so that during (1) there is another component implementing the behaviour of (2). That said, such a component would be subject to the same limitations as the agent for implementing policy for newly learned DNS names. Longer term the full solution is probably to push more DNS understanding down into the datapath in order to allow it to handle more dynamic name<->IP mapping updates, but I'm mostly speculating about capabilities there. Designing such a solution is a whole thing of its own.

christarazi replied:
I'm hoping things like https://github.com/cilium/cilium/pull/27163 minimize the impact for high-scale clusters.

squeed replied:
This PR doesn't change the length of state (1); the DNS proxy opens its socket same as always. On large clusters, stage (2) is definitely longer. I think we can be somewhat more selective as to what blocks non-cluster-entity injection. We don't need, say, nodes or endpoints to be synced. That's why I think there's room to improve this.

christarazi replied:
Do we believe that making stage (2) longer is acceptable or do we need to block this PR?

squeed replied:
Chatted with @christarazi; the summary is:

- This PR definitely has a behavior change, we should consider documenting it in the release notes
- DNS updates that return a new IP may cause drops until the following set of resources are synchronized:
  - Service
  - Namespace
  - Pod (local pods only)
  - Node
  - EndpointSlices
  - NetworkPolicy
  - Secret (single namespace only)
  - CiliumNetworkPolicy
  - CiliumClusterwideNetworkPolicy
  - CiliumEndpoint
  - CiliumNode
  - CiliumLocalRedirectPolicy
  - CiliumEndpointSlice
  - CiliumCIDRGroup
 
Chris' suggestion is to work on improving startup time generically, which I agree is a good way forward. There are ongoing efforts to improve policy ingestion (viz #27163 and others). My proposal to skip certain resources has merit, but it causes a fragile and tight coupling between wildly disparate parts of the codebase, and should be avoided.

So, I suppose the only "resolution" is to either decide the pause in state (2) is acceptable, or decide that identity flaps are acceptable. (I have some work on handling identity flaps robustly, but it's not ready to merge).

joestringer replied:
Identity flaps scare me a bit. The less churn we have there, the better. In particular if we are renumbering many identities at once, then during the renumbering period we could drastically increase pressure on policymaps and in worst case bump over policymap size limits.

My initial thinking was that if we defer starting the dns proxy until the syncing occurs, it would increase (1) and that's kinda better logically just because the agent can only handle so many requests in state (2) anyway and they're all getting blocked for a certain time period. But I think the differences between (1) and (2) in those cases is relatively small, so between those options I'm fine with the status quo here.

As for the behaviour change and release-noting it, I'm inclined to agree, though I don't have a good idea on how to communicate the change effectively. I feel like in the grand scheme of things this is a push in the direction of correctness and properly synchronizing ipcache state updates between multiple data sources, which will lead to better visibility in future as well as reducing the likelihood of packet drops due to identity churn / flapping. On the other hand it can delay how long it takes for Cilium to be ready to plumb new policy for `toFQDNs` policies after startup, particularly in larger clusters.

squeed replied:
I should have tested this days ago. Apologies for the much ado about nothing.

Conclusion: In a caches-unsynched agent (i.e. case (2), we *still can't plumb new responses*, because there are no selectors, because endpoints haven't regenerated yet. Because there are no selectors, there is no day in propagating selector changes.

So this PR neither regresses on behavior nor adds an additional delay. This conversation can be resolved, and I learned a valuable lesson: test! before! worrying!

joestringer commented on daemon/cmd/fqdn.go:
```diff
@@ -64,112 +59,74 @@ const (
+			logger.Debug("FQDN selector update requires IPCache completion.")
+			<-ipcacheDoneChan
+			wg.Done()
+		}()
```
Speaking also a bit to [Chris' earlier comment](https://github.com/cilium/cilium/pull/29036#discussion_r1395136606), this is a bit awkward to have the `IPCache` return a channel and then this code needs to then spawn a goroutine (per DNS response) in order to translate this into a WaitGroup. But then, if we look another couple of layers up the call stack into `notifyOnDNSMsg()` we'll find _another_ goroutine converting this WaitGroup back into a channel in order to do a `select {}` on this and another timeout.

Functionally I think this is correct, but it seems like this code needs a bit of an overhaul in terms of WaitGroup vs. channel handling.

squeed replied:
This is all such a spiderweb, there are at least 3 WaitGroups ultimately in use, plus a channel for timeout purposes.

I don't think we can get away from using waitgroups as the primary synchronization point (without a painful refactor), but I did eliminate the channel by switching to a revision number.

I'm not sure if there's much else to be done here. Perhaps we can make a convenience package for time-out-able waitgroups.

christarazi replied:
> Perhaps we can make a convenience package for time-out-able waitgroups.

FWIW, we do have `StoppableWaitGroup` in the code which might have the characteristics you need.

squeed replied:
Good point, converted to StoppableWaitGroup.

squeed replied:
Oh right, I didn't do this before because StoppableWaitGroup has some weird semantics when the value of the counter is zero.

joestringer commented on daemon/cmd/daemon.go:
nit: In order to clarify that ipcache is now a hard dependency, it may be worth propagating this dependency up to `newDaemon()` by adding it as an argument to `bootstrapFQDN()`. This way, if/when we further refactor initialization, this ordering requirement will be more clear. Could save us a bit of pain in future.

```
diff --git a/daemon/cmd/daemon.go b/daemon/cmd/daemon.go
index a7b84efd2fab..a399b6a6e2dc 100644
--- a/daemon/cmd/daemon.go
+++ b/daemon/cmd/daemon.go
@@ -680,8 +680,9 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
        }
        bootstrapStats.restore.End(true)

        bootstrapStats.fqdn.Start()
-       err = d.bootstrapFQDN(restoredEndpoints.possible, option.Config.ToFQDNsPreCache)
+       err = d.bootstrapFQDN(restoredEndpoints.possible, option.Config.ToFQDNsPreCache, d.ipcache)
        if err != nil {
                bootstrapStats.fqdn.EndError(err)
                return nil, restoredEndpoints, err
```

joestringer commented on daemon/cmd/fqdn.go:
```diff
@@ -64,112 +59,74 @@ const (
-
-	return selectorIdentitySliceMapping, usedIdentities, newlyAllocatedIdentities, nil
-}
+// used as a context key to extract
```
nit: I don't understand this comment at all. Could be improved or removed.
```suggestion
```

squeed replied:
It's a key to `context.Value()`. I've updated the comment.

joestringer commented on pkg/fqdn/cache.go:
```diff
@@ -560,6 +562,21 @@ func (c *DNSCache) removeReverse(ip netip.Addr, entry *cacheEntry) {
 }
 
+// dumpReverse takes a snapshot of all IPs in the reverse cache.
+// It it used for un-allocating IP addresses.
```
nit: It's bad form to make assumptions about callers, as the callers can be refactored and it makes these comments wrong. It's pretty hard to keep such comments up to date. If we want to know that it's used for un-allocating IP addresses, we can look at the callers and clearly see that this is how it's used.
```suggestion
```

If there's a property that this call provides which is relevant to the fact that it's used for un-allocating IP addresses, such as taking a copy rather than a reference of underlying state, then that would be more reasonable to include in the func doc here.

squeed replied:
Very fair point, agreed.

joestringer commented on pkg/fqdn/name_manager_gc.go:
```diff
@@ -0,0 +1,269 @@
+// This implements some garbage cllection and cleanup functions for the NameManager
+
+// GC is a controller to cleanup TTL expired entries from the DNS policies.
+// dns-garbage-collector-job runs the logic to remove stale or undesired
```
nit:
```suggestion
// This file implements some garbage collection and cleanup functions for the NameManager

// GC removes stale or undesired
```

(GC is no longer a controller after refactor, code does not match the function comment)

joestringer commented on pkg/fqdn/name_manager_gc.go:
```diff
@@ -0,0 +1,269 @@
+		}
+	}
+
+	// Ensure there's a metadata entry for tne fqdn subsystem in ipcache.
```
nit:
```suggestion
	// Ensure there's a metadata entry for the fqdn subsystem in ipcache.
```

joestringer commented on pkg/fqdn/name_manager_test.go:
```diff
@@ -24,63 +26,62 @@ import (
-			Cache:  NewDNSCache(0),
+			MinTTL:  1,
+			Cache:   NewDNSCache(0),
+			IPCache: &fakeIPCache{},
```
nit: If you replace this with the MockIPCache, no need to declare a new one in this package.

The Upsert/Remove prefixes calls in the fake IPCache were also unused so far. Not sure if you intended to use those as a basis for some subsequent unit testing.

```
diff --git a/pkg/fqdn/name_manager_test.go b/pkg/fqdn/name_manager_test.go
index 7b9d6cd098f0..1273fba4580d 100644
--- a/pkg/fqdn/name_manager_test.go
+++ b/pkg/fqdn/name_manager_test.go
@@ -14,9 +14,8 @@ import (

        "github.com/cilium/cilium/pkg/checker"
        "github.com/cilium/cilium/pkg/fqdn/dns"
-       ipcacheTypes "github.com/cilium/cilium/pkg/ipcache/types"
        "github.com/cilium/cilium/pkg/policy/api"
-       "github.com/cilium/cilium/pkg/source"
+       testipcache "github.com/cilium/cilium/pkg/testutils/ipcache"
 )

 // TestNameManagerCIDRGeneration tests rule generation output:
@@ -31,7 +30,7 @@ func (ds *FQDNTestSuite) TestNameManagerCIDRGeneration(c *C) {
                nameManager = NewNameManager(Config{
                        MinTTL:  1,
                        Cache:   NewDNSCache(0),
-                       IPCache: &fakeIPCache{},
+                       IPCache: testipcache.NewMockIPCache(),

                        UpdateSelectors: func(ctx context.Context, selectorIPMapping map[api.FQDNSelector][]netip.Addr, _ <-chan struct{}) *sync.WaitGroup {
                                for k, v := range selectorIPMapping {
@@ -75,7 +74,7 @@ func (ds *FQDNTestSuite) TestNameManagerMultiIPUpdate(c *C) {
                nameManager = NewNameManager(Config{
                        MinTTL:  1,
                        Cache:   NewDNSCache(0),
-                       IPCache: &fakeIPCache{},
+                       IPCache: testipcache.NewMockIPCache(),

                        UpdateSelectors: func(ctx context.Context, selectorIPMapping map[api.FQDNSelector][]netip.Addr, _ <-chan struct{}) *sync.WaitGroup {
                                for k, v := range selectorIPMapping {
@@ -138,28 +137,3 @@ func (ds *FQDNTestSuite) TestNameManagerMultiIPUpdate(c *C) {
        nameManager.Unlock()

 }
-
-type fakeIPCache struct {
-       prefixes map[netip.Prefix]struct{}
-}
-
-func (i *fakeIPCache) UpsertPrefixes(prefixes []netip.Prefix, src source.Source, resource ipcacheTypes.ResourceID) <-chan struct{} {
-       if i.prefixes == nil {
-               i.prefixes = make(map[netip.Prefix]struct{})
-       }
-       for _, p := range prefixes {
-               i.prefixes[p] = struct{}{}
-       }
-       out := make(chan struct{})
-       close(out)
-       return out
-}
-
-func (i *fakeIPCache) RemovePrefixes(prefixes []netip.Prefix, src source.Source, resource ipcacheTypes.ResourceID) {
-       if i.prefixes == nil {
-               return
-       }
-       for _, p := range prefixes {
-               delete(i.prefixes, p)
-       }
-}
diff --git a/pkg/testutils/ipcache/ipcache.go b/pkg/testutils/ipcache/ipcache.go
index 89922bf5763f..3174794c1cb6 100644
--- a/pkg/testutils/ipcache/ipcache.go
+++ b/pkg/testutils/ipcache/ipcache.go
@@ -54,3 +54,12 @@ func (m *MockIPCache) DeleteOnMetadataMatch(IP string, source source.Source, nam
 func NewMockIPCache() *MockIPCache {
        return &MockIPCache{}
 }
+
+func (m *MockIPCache) UpsertPrefixes(prefixes []netip.Prefix, src source.Source, resource ipcacheTypes.ResourceID) <-chan struct{} {
+       out := make(chan struct{})
+       close(out)
+       return out
+}
+
+func (m *MockIPCache) RemovePrefixes(prefixes []netip.Prefix, src source.Source, resource ipcacheTypes.ResourceID) {
+}
```

squeed replied:
Oh, good spot.

joestringer commented on pkg/ipcache/metadata.go:
```diff
@@ -80,6 +80,13 @@ type metadata struct {
+	// might fail, in which case we re-queue the set of pending prefixes
+	// but have already allocated a new done channel.
+	// The length of this is theoretically unbound, but in reality it will
+	// never have more than 2 entries. It will always have at least one entry.
```
I think that it's helpful to frame this as "expected to be one entry long" for the reader, but documenting that it will not have more than 2 entries is potentially misleading here at the field. Reading the implementation, I can't tell where the 2 entry bounding is. If the ipcache gets full and the controller continues to trigger, it looks like this will continue to grow unbounded.

I'd suggest (1) documenting at the field the expectations and not assertions about the users, then (2) document at the users how we bound the update of this field, so that the code controlling the bounding is right next to the comment that asserts the bounding.

```suggestion
	// The length of this is at least one entry, and it typically contains one entry.
```

squeed replied:
I've since removed this code and switched to a revision number at @christarazi's suggestion.

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -695,132 +248,103 @@ func (l *labelIdentitySelector) fetchIdentityMappings() []identity.NumericIdenti
+//
+// Returns an UpdateResult that indicates if the caller should call UpdatePolicyMaps() and/or
+// wait for ipcache UpsertMetadata() to finish.
+func (sc *SelectorCache) UpdateFQDNSelector(fqdnSelec api.FQDNSelector, ips []netip.Addr, wg *sync.WaitGroup) UpdateResult {
```
nit: This function comment is now incomplete, could do with an update. A few notes:

- This function will update the selector's mapping for IP addresses.
- It will also enqueue policymap updates for endpoints iff the IP is already used by another policy.
- This function will *not* initiate allocation of new identities, nor policymap updates for new identities, nor ipcache updates for new identities.
- Any IPs that were not previously allocated via another mechanism (such as another network policy or a previous DNS lookup) must be separately plumbed by callers into the ipcache, which will take care of ensuring updates to the datapath (including policymaps).

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -695,132 +248,103 @@ func (l *labelIdentitySelector) fetchIdentityMappings() []identity.NumericIdenti
-		fqdnSel = selector.(*fqdnSelector)
+	idSelector, exists := sc.selectors[key]
+	if !exists || idSelector == nil {
+		log.WithField(logfields.Selector, fqdnSelec.String()).Error("UpdateFQDNSelector of selector not registered in SelectorCache!")
```
Did something functionally change about the control flow such that this path no longer initializes new selectors? Or was the creation of a selector from this path previously just dead code that was never used?

If we have good paths to exercise this code then I'm not against this, but I'm not sure whether this path is being fully reasoned through. For instance, if the selector is not cached then why would this function continue and will that have unforeseen effects?

joestringer replied:
Ah, I see that in the final patch this case should probably return due to the failure to cast to an `fqdnSelector`, so probably this is now fine... though I still don't fully understand why the previous code would create a new FQDN selector here.

squeed replied:
This is the same situation as https://github.com/cilium/cilium/pull/29036#discussion_r1406029335 -- because we had to unlock the selector cache during AddFQDNSelector, it was possible for updates to be received before the selector add was complete. This is no longer possible.

joestringer commented on pkg/fqdn/helpers.go:
```diff
@@ -16,15 +18,14 @@ import (
 // FQDNSelector. Returns the mapping of DNSName to set of IPs which back said
 // DNS name, the set of FQDNSelectors which do not map to any IPs, and the set
 // of FQDNSelectors mapping to a set of IPs.
-func (n *NameManager) MapSelectorsToIPsLocked(fqdnSelectors map[api.FQDNSelector]struct{}) (selectorsMissingIPs []api.FQDNSelector, selectorIPMapping map[api.FQDNSelector][]net.IP) {
```
nit: Function comment needs updating for parameter changes.

joestringer commented on pkg/fqdn/helpers.go:
```diff
@@ -36,14 +37,6 @@ func (n *NameManager) MapSelectorsToIPsLocked(fqdnSelectors map[api.FQDNSelector
 			dnsName := prepareMatchName(ToFQDN.MatchName)
 			lookupIPs := n.cache.Lookup(dnsName)
 
-			// Mark this FQDNSelector as having no IPs corresponding to it.
-			// FQDNSelectors are guaranteed to have only their MatchName OR
-			// their MatchPattern set (having both set is invalid per
-			// sanitization of FQDNSelectors).
-			if len(lookupIPs) == 0 {
-				missing[ToFQDN] = struct{}{}
-			}
```
I couldn't quite tell at a glance, but what's the reason why the new implementation no longer needs to know which selectors lack IPs?

squeed replied:
Good question. It was used as an "edge" to detect selectors for which identities can be deallocated. All of this goes away now that selectors no longer own identities.

joestringer commented on pkg/fqdn/name_manager.go:
```diff
@@ -64,47 +69,54 @@ func (n *NameManager) GetModel() *models.NameManager {
+// UnregisterForIPUpdatesLocked.
+// Because the NameManager and SelectorCache have interleaving locks, we
+// must ALWAYS lock the NameManager first before locking the SelectorCache.
+// Hence, this must be made available.
```
nit, this line adds no context.
```suggestion
```

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -695,132 +248,103 @@ func (l *labelIdentitySelector) fetchIdentityMappings() []identity.NumericIdenti
+	// UpdateResultUpdatePolicyMaps indicates the caller should call EndpointManager.UpdatePolicyMaps()
+	UpdateResultUpdatePolicyMaps = UpdateResult(1) << iota
+
+	// UpdateResultIdentitiesNeeded indicates the caller should wait for a complete ipcace label injection,
```
nit:
```suggestion
	// UpdateResultIdentitiesNeeded indicates the caller should wait for a complete ipcache label injection,
```

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -695,132 +248,103 @@ func (l *labelIdentitySelector) fetchIdentityMappings() []identity.NumericIdenti
+	// Delete any non-matching entries from cachedSelections
+	for nid := range idSelector.cachedSelections {
+		identity := sc.idCache[nid]
+		if !idSelector.source.matches(identity) {
```
I'm idly wondering whether the new path with more `matches()` calls may be more inefficient. Not sure whether this path is well set up for some benchmarking.

squeed replied:
Overall there is a bit more work within the SelectorCache, since we now scan all IDs on adding a FQDNSelector. However, we don't have to perform allocation, so that is probably a net win. In any case, FQDNSelectors and LabelIdentitySelectors now look basically the same, so that is a nice little cleanup.

This particular change doesn't actually add any more matches, it's just an interface cleanup.

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -829,78 +353,60 @@ func (sc *SelectorCache) updateFQDNSelector(fqdnSelec api.FQDNSelector, identiti
-	// 'ids' slice. Either way, they are added to the selector that is
-	// stored in 'sc.selectors[]'.
-	currentlyAllocatedIdentities := sc.allocateIdentityMappings(fqdnSelec, selectorIPMapping)
+	currentIPs := sc.localIdentityNotifier.RegisterForIPUpdatesLocked(source.selector)
```
Do you understand the previous "If this was called twice" arguments? Do you know how they now apply with the new logic? I'm not sure the significance of this, but given that the comment is being removed, I'm also not sure whether the complexity goes away because perhaps we're switching to tracking IPs with labels (stateless) vs. identities (stateful).

squeed replied:
It's even simpler than that -- in the previous version of the code, the NameManager might need to allocate identities during policy calculation. In order for that to succeed, the SelectorCache had to unlock itself first. But, because the SelectorCache was unlocked, it could theoretically get another `AddFQDNPolicy()` of the _same policy_. 

In this version of the code, locks are no longer dropped and it is not possible for a policy to be double-registered. So, we can remove the heroic code that dealt with this situation.

joestringer reviewed (approved):
It's awesome to see removal of a bunch of the complex identity reference counting logic in this PR. I'm also glad to see the FQDNSelector start to conform more closely with the way that the identitySelector works internally in the selectorcache. Those diverging paths have previously concerned me a bit in the duplicated codepaths to handle them and potential for code misalignment if we don't keep them equally up to date.

I've commented on the main concern you raised as a thread below and pulled in Chris' comments. Hopefully having a single regular thread to go back and forth on should make it easier to reason through that concern.

Reviewing my review, I think I'm not proposing to change anything significant at all. There's a couple of discussion threads but most of my comments are just nits that I observed while laying eyes on all of the changes. We could certainly debate a bit about the exact synchronization mechanism for ipcache, but in terms of moving the code in a positive direction without significant flaws, I think that this PR passes muster as-is.

squeed committed:
ipcache: add ipcache revision counter

This is a way for ipcache sources to wait for queued updates to be
applied. The Update* calls return a revision number, which can then be
passed to `ipcache.WaitForRevision()`. Upon a successful injection, the
revision is updated and waiters are triggered.

Internally, it uses a sync.Cond{} to broadcast updates.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

christarazi reviewed (approved):
PR looks good to me as-is. New revision implementation makes sense. There are a couple of outstanding conversations left to be resolved which still require some consensus to be reached.

joestringer commented on daemon/cmd/policy.go:
```diff
@@ -125,6 +125,7 @@ func newPolicyTrifecta(params policyParams) (policyOut, error) {
 
 	params.Lifecycle.Append(hive.Hook{
 		OnStart: func(hc hive.HookContext) error {
+			ipc.Start()
```
Cilium can be very sensitive in startup ordering, how does this change impact the initial trigger of IPCache state synchronization?

Separately, it seems like `newPolicyTrifecta` is used in tests, so will running a controller there have other unintended follow-on effects?

squeed replied:
(removed this commit)

joestringer commented on daemon/cmd/policy.go:
The commit message for this first commit also alleges buggy behaviour without details. Does that mean we should be splitting this commit into a separate PR and backporting it? Or investigating that buggy behaviour?

squeed replied:
(removed this commit)

squeed replied:
yes, we should be backporting this.

joestringer commented on pkg/ipcache/metadata.go:
```diff
@@ -658,26 +700,28 @@ func (ipc *IPCache) TriggerLabelInjection() {
+	if !ipc.Configuration.CacheStatus.Synchronized() {
+		return errors.New("k8s cache not fully synced")
+	}
+
```
Why duplicate them? This is the only functional user.

squeed replied:
Removed this whole commit.

joestringer commented on daemon/cmd/fqdn.go:
I skipped the ~500 lines of patch that made significant changes in this file because it's too big. I'll take your word that there's no functional changes.

joestringer replied:
Honestly refactoring like that should just be a PR by itself. This PR is too big.

squeed replied:
You’re right, and this snowballed a bit after adding the revision mechanism.

The bit refactoring the ipcache initialization is not actually required (I had the commit lying around). I’ll pull it out and move it to a separate PR.

joestringer reviewed (commented):
On second pass I mainly reviewed the new revision approach for syncing with ipcache to ensure updates are propagated to the datapath. I like it quite a lot.

I left a couple of comments about startup ordering and concerns about this PR potentially addressing bugs that will then only be present on 1.13/1.14.

I don't love the size of this PR.

tklauser commented on pkg/ipcache/metadata_test.go:
```diff
@@ -771,6 +771,49 @@ func TestRequestIdentity(t *testing.T) {
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		m.waitForRevision(wantRev)
+		wg.Done()
+	}()
```
nit: should we add a timeout here in case some future change were to break `waitForRevision` to wait indefinitely? I guess it will be caught by the global test timeout but a shorter timeout might be worthwhile.

squeed replied:
That would add an extra goroutine and channel. Given that the waitgroup in question already has a timeout, that's probably not worth it.

squeed replied:
(This is a "leaf" operation and nothing else waits on it, so there's no benefit in ending early).

tklauser replied:
> (This is a "leaf" operation and nothing else waits on it, so there's no benefit in ending early).

:facepalm: nevermind

tklauser commented on pkg/ipcache/metadata.go:
```diff
@@ -94,30 +106,61 @@ func newMetadata() *metadata {
 }
 
-func (m *metadata) enqueuePrefixUpdates(prefixes ...netip.Prefix) {
+// enqueuePrefixUpdates queues prefixes for label injection. It returns the
```
This godoc comment seems incomplete.

tklauser commented on pkg/ipcache/metadata.go:
```diff
@@ -94,30 +106,61 @@ func newMetadata() *metadata {
+// supplied revision. We may skip revisions, as the desired revision is bumped
+// every time prefixes are dequeued, but injection may fail. Thus, any revision
+// greater or equal to the desired revision is acceptable.
+func (m *metadata) waitForRevision(rev uint64) {
```
nit: I wonder whether we should name this `waitForRevisionGreaterEqual` or similar to make it clear that it's not waiting for the exact revision, also considering it's mentioned twice in the godoc :)

squeed replied:
I thought about that, but decided that it was too many words :-).

tklauser commented on pkg/ipcache/ipcache.go:
```diff
@@ -589,6 +598,13 @@ func (ipc *IPCache) RemoveIdentityOverride(cidr netip.Prefix, identityLabels lab
 	ipc.RemoveMetadata(cidr, resource, overrideIdentity(true), identityLabels)
 }
 
+// WaitForRevision will block until the desired revision has been reached. It can be used
```
Should we move/copy the `(*metadata).waitForRevision` godoc here? I think it's worth mentioning in the public API that it's waiting for the revision to be ≥ the desired revision. Also might be worth to rename the method to `WaitForRevisionGreaterEqual` (ref. my review comment on `(*metadata).waitForRevision`.

squeed replied:
Tightened up the wording a bit.

tklauser commented on pkg/ipcache/metadata.go:
```diff
@@ -81,6 +81,18 @@ type metadata struct {
 	queuedPrefixes  map[netip.Prefix]struct{}
 
+	// queuedRevision is the "version" of the prefix queue. It is incremented
+	// on every *dequeue*. If injection is successful, then injectedRevision
```
It seems weird that the godoc comment for `queuedRevision` describes when `injectedRevision` is updated. Should the second sentence be moved to the `injectedRevision` godoc?

squeed replied:
Just trying to provide additional context, but I can try and make this clearer.

tklauser commented on pkg/ipcache/metadata.go:
```diff
@@ -94,30 +106,61 @@ func newMetadata() *metadata {
 		queuedPrefixes: make(map[netip.Prefix]struct{}),
+		queuedRevision: 1,
+
+		injectedRevisionCond: &sync.Cond{L: &lock.Mutex{}},
```
nit: 
```suggestion
		injectedRevisionCond: sync.NewCond(&lock.Mutex{}},
```

tklauser commented on pkg/fqdn/name_manager_gc.go:
```diff
@@ -0,0 +1,260 @@
+const DNSGCJobInterval = 1 * time.Minute
+
+var (
+	dnsGCJobName         = "dns-garbage-collector-job"
```
nit: this can be a const

squeed replied:
Somehow I was convinced strings couldn't be `const`. Huh.

tklauser commented on pkg/fqdn/name_manager_gc.go:
```diff
@@ -0,0 +1,270 @@
+	dnsGCControllerGroup = controller.NewGroup(dnsGCJobName)
+)
+
+// This implements some garbage cllection and cleanup functions for the NameManager
```
```suggestion
// This implements garbage collection and cleanup functions for the NameManager.
```

tklauser commented on pkg/fqdn/name_manager_gc.go:
```diff
@@ -40,17 +42,16 @@ var (
-		// with the CT GC.
-		activeConnectionsTTL = int(2 * DNSGCJobInterval.Seconds())
-		activeConnections    = NewDNSCache(activeConnectionsTTL)
-	)
```
This part of the change looks unrelated to commit `fqdn: (cleanup) use sets.Set instead of maps`. It should probably be part of the previous commit (`fqdn: move GC, cleanup, restore functions to NameManager`).

squeed replied:
Ooh, good catch.

tklauser commented on pkg/fqdn/name_manager_gc.go:
```diff
@@ -133,6 +137,13 @@ func (n *NameManager) GC(ctx context.Context) error {
+	// Need to RLock here so we don't race on re-insertion.
+	n.RWMutex.RLock()
+	n.maybeRemoveMetadataLocked(maybeStaleIPs)
+	n.RWMutex.RUnlock()
```
It seems both callers of `maybeRemoveMetadataLocked` look like that. Maybe move the read lock and the comment into `maybeRemoveMetadataLocked` so not to obstruct the main control flow with details about locking?

squeed replied:
Oh, good point.

tklauser commented on daemon/cmd/fqdn.go:
```diff
@@ -64,112 +59,75 @@ const (
+type requestNameKey struct{}
 
-func (d *Daemon) updateSelectorCacheFQDNs(ctx context.Context, selectors map[policyApi.FQDNSelector][]*identity.Identity, selectorsWithoutIPs []policyApi.FQDNSelector) *sync.WaitGroup {
+// updateSelectors propagates the mapping of FQDNSelector to IP
```
nit:

```suggestion
// updateSelectors propagates the mapping of FQDNSelector to IPs
```

tklauser commented on pkg/fqdn/name_manager.go:
```diff
@@ -270,12 +270,12 @@ func (n *NameManager) updateIPsForName(lookupTime time.Time, dnsName string, new
 var ipcacheResource = ipcacheTypes.NewResourceID(ipcacheTypes.ResourceKindDaemon, "", "fqdn-name-manager")
 
 // upsertMetadata adds an entry in the ipcache metadata layer for the set of IPs
```
nit: document the return value

tklauser commented on daemon/cmd/daemon_main.go:
```diff
@@ -878,11 +878,11 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.Int(option.DNSProxyLockCount, 131, "Array size containing mutexes which protect against parallel handling of DNS response IPs. Preferably use prime numbers")
-	flags.MarkHidden(option.DNSProxyLockCount)
+	flags.MarkDeprecated(option.DNSProxyLockCount, "This option no longer has any effect, and will be removed in v1.16")
 	option.BindEnv(vp, option.DNSProxyLockCount)
 
 	flags.Duration(option.DNSProxyLockTimeout, 500*time.Millisecond, fmt.Sprintf("Timeout when acquiring the locks controlled by --%s", option.DNSProxyLockCount))
-	flags.MarkHidden(option.DNSProxyLockTimeout)
+	flags.MarkDeprecated(option.DNSProxyLockTimeout, "This option no longer has any effect, and will be removed in v1.16")
```
Please also add a note to the respective section in `Documentation/operations/upgrade.rst` mentioning that these options are deprecated.

tklauser commented on pkg/policy/selectorcache_selector.go:
```diff
@@ -0,0 +1,358 @@
+// cases below identitySelector is being implemented by structs.
+//
+// Because identitySelector is exposed to the user is used as a map key, it must always
+// be passed to the user as a pointer to the actual type.
```
I think this no longer applies now that `identitySelector` is no longer an interface, ref. d84ee5215c7a695a8c23bcfa41098d8860de58e5. Though it's probably still beneficial to pass as a pointer for different reasons.

squeed replied:
It does still apply, since the only exposed type is the CachedSelection interface. I'll clarify this in the doc.

tklauser reviewed (commented):
Did a first pass focusing mainly on godoc clarity and code readability. Will do a second pass focusing more on other aspects (e.g. potential scalability issues) once unresolved review comments from Chris and Joe are addresses.

I like that in total this removes more than 200 LoC :smiley:

tklauser commented on Documentation/operations/upgrade.rst:
```diff
@@ -375,6 +375,9 @@ Removed Options
 * The long defunct and undocumented ``single-cluster-route`` flag has been removed.
 
+* The ``dnsproxy-lock-count`` and ``dnsproxy-lock-timeout`` flags no longer have
+  any effect and are deprecated. They will be removed in a future release.
```
nit: 

```suggestion
  any effect and are deprecated. They will be removed in v1.16.
```

given that the option usage says so :smile:

squeed replied:
This PR is green :-) I'll fix this separately :stuck_out_tongue_winking_eye:

squeed replied:
https://github.com/cilium/cilium/pull/29531

tklauser reviewed (approved):
My review comments have been addressed. Didn't have time for an in-depth 2nd pass, but I've briefly skimmed through the changes again and reading up on the thread with Chris and Joe this looks OK to me.

One very minor nit in case another respin/rebase is needed (I guess you anyway need a @cilium/docs-structure review for that part of the change).

joestringer reviewed (approved):
Approving the docs change, subject to Tobias' comment.
